cellarium/ml/layers/transformer.py

```diff
@@ -0,0 +1,196 @@
+    """
+    Transformer model.
+
+    Args:
+        d_model:
+            Dimensionality of the embeddings and hidden states.
+        d_ffn:
+            Dimensionality of the inner feed-forward layers.
+        use_bias:
+            Whether to use bias in the linear transformations and norm-add layers.
+        n_heads:
+            Number of attention heads.
+        n_blocks:
+            Number of transformer blocks.
+        dropout_p:
+            Dropout probability.
+        attention_logits_scale:
+            Multiplier for the attention scores.
+        attention_backend:
+            Backend for the attention computation.
+        Wqkv_initializer:
+            Initializer for the query, key, and value linear transformations.
+        Wo_initializer:
+            Initializer for the output linear transformation.
+        dense1_initializer:
+            Initializer for the first dense layer.
+        dense2_initializer:
+            Initializer for the second dense layer.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        d_ffn: int,
+        use_bias: bool,
+        n_heads: int,
+        n_blocks: int,
+        dropout_p: float,
+        attention_logits_scale: float,
+        attention_backend: Literal["math", "flash", "mem_efficient", "torch"],
+        attention_softmax_fp32: bool,
+        Wqkv_initializer: dict[str, Any],
+        Wo_initializer: dict[str, Any],
+        dense1_initializer: dict[str, Any],
+        dense2_initializer: dict[str, Any],
+    ) -> None:
+        super().__init__()
+        self.n_blocks = n_blocks
+        self.blocks = nn.ModuleList(
+            [
+                TransformerBlock(
+                    d_model,
+                    d_ffn,
+                    use_bias,
+                    n_heads,
+                    dropout_p,
+                    attention_logits_scale,
+                    attention_backend,
+                    attention_softmax_fp32,
+                    Wqkv_initializer,
+                    Wo_initializer,
+                    dense1_initializer,
+                    dense2_initializer,
+                )
+                for _ in range(n_blocks)
+            ]
+        )
+        self.ln = nn.LayerNorm(d_model, use_bias)
+
+        self._reset_parameters()
+
+    def _reset_parameters(self) -> None:
+        self.ln.reset_parameters()
+
+    def forward(
+        self,
+        hidden_state_ncd: torch.Tensor,
+        attention_mask_ncc: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Args:
+            hidden_state_ncd:
+                Hidden state tensor of shape ``(n, c, d)``.
+            attention_mask_ncc:
+                Attention mask tensor of shape ``(n, c, c)``.
+
+        Returns:
+            The output hidden state tensor of shape ``(n, c, d)``.
+        """
+        for block in self.blocks:
+            hidden_state_ncd = block(hidden_state_ncd, attention_mask_ncc)
+
+        return self.ln(hidden_state_ncd)
```

Comment:
Added this final LN which seems to be common in pre-LN transformer architectures. I see improvements in training with it judging from the training loss curves and grad norm curves.

w/ final LN
<img width="345" alt="Screenshot 2024-10-13 at 10 36 29 AM" src="https://github.com/user-attachments/assets/b1a802ba-48ef-4c6a-a159-7d59b28d7fc6">
<img width="346" alt="Screenshot 2024-10-13 at 10 36 41 AM" src="https://github.com/user-attachments/assets/6f26316d-493b-45db-a809-7dcb97c62b5d">
<img width="343" alt="Screenshot 2024-10-13 at 10 36 52 AM" src="https://github.com/user-attachments/assets/4a762c5c-76d6-4029-ad6d-205dce96499d">

w/o final LN
<img width="344" alt="Screenshot 2024-10-13 at 10 37 27 AM" src="https://github.com/user-attachments/assets/770ac6f8-255a-481b-bd81-50eec907569d">
<img width="347" alt="Screenshot 2024-10-13 at 10 37 34 AM" src="https://github.com/user-attachments/assets/96873ea3-fbc8-4f4d-aaa7-1d354bb01745">
<img width="343" alt="Screenshot 2024-10-13 at 10 37 43 AM" src="https://github.com/user-attachments/assets/1601cc12-163d-48ec-8abd-3aefad050a5b">